python/python/Ice/Future.py

```diff
@@ -1,24 +1,193 @@
+
+        If the future was cancelled before completing, an `InvocationCanceledException` is raised.
+
+        If the operation completed successfully without raising an exception, `None` is returned.
+
+        Parameters
+        ----------
+        timeout : int or float, optional
+            Maximum time (in seconds) to wait for the exception. If `None`, the method waits indefinitely until the
+            operation completes.
+
+        Returns
+        -------
+        BaseException or None
+            The exception raised by the operation, or `None` if the operation completed successfully.
+
+        Raises
+        ------
+        TimeoutException
+            If the operation has not completed within the specified timeout.
+        InvocationCanceledException
+            If the operation was cancelled before completing.
+        """
+
+    @abstractmethod
+    def set_exception(self, ex: BaseException):
+        """
+        Set an exception for the future and mark it as completed.
+
+        This method stores the provided exception `ex` and transitions the future's state to "done". Any registered
+        callbacks are executed after the state update.
+
+        If the future is not in a running state, this method has no effect.
+
+        Parameters
+        ----------
+        ex : Exception
+            The exception to store in the future.
+        """
+        pass
+
+    @abstractmethod
+    def cancel(self) -> bool:
+        """
+        Attempt to cancel the operation.
+
+        If the operation is already running or has completed, it cannot be cancelled, and the method returns False.
+        Otherwise, the operation is cancelled, and the method returns True.
+
+        Returns
+        -------
+        bool
+            True if the operation was cancelled, False otherwise.
+        """
+        pass
+
+    @abstractmethod
+    def cancelled(self) -> bool:
+        """
+        Check if the future has been cancelled.
+
+        Returns
+        -------
+        bool
+            True if the future was cancelled using `cancel()`, otherwise False.
+        """
+        pass
+
+    @abstractmethod
+    def add_done_callback(self, fn: Callable) -> None:
+        """
+        Attach a callback to be executed when the future completes.
+
+        The callback `fn` is called with the future as its only argument once the future completes or is cancelled.
+        Registered callbacks are executed in the order they were added.
+
+        If the future is already complete, `fn` is called immediately from the calling thread.
+
+        Parameters
+        ----------
+        fn : callable
+            The function to execute upon completion. Takes the future as its argument.
+        *args : Any
+            Additional positional arguments (for compatibility with different Future implementations).
+        **kwargs : Any
+            Additional keyword arguments (for compatibility with different Future implementations).
```

Comment:
Are `*args` and `**kwargs` parameters for `add_done_callback`?

Reply:
No, I was trying stuff earlier. I caught that earlier and removed.